### DISCUSSION TITLE
$persist - an option to save a variable value to the Session Storage (instead of the Local Storage)

### DISCUSSION THREAD

**Discussion by stepanjakl** [alpinejs/alpine#1988] [3. Feature ideas]
Hello,

Thank you for an amazing library and additional tools such as the Persist helper.

The Persist plugin allows to save the values to the Local Storage. It may be worth giving an option to save into the Session Storage. I see a use case in the use of page banners which you may want to show every time the user comes back. A key message/news could change and the Local Storage variable would hide the banner if the user dismissed it in the previous session(s). I believe that Local Storage has no expiration time, whereas Session Storage clears all data at the end of the page session.

Syntax could look like this for instance: `$persist.session(var)` or `$persist(var).via('sessionStorage')` /  `$persist.indefinitely(var)` or `$persist.via('localStorage')` ETC. (other suggestions are welcomed)

What do you think?:)

PS: eventually cookies could be included with expiry time periods as well

**Comment by SimoTod** [comment]
The only problem with sessionStorage is that it doesn't work as people expect. The session is per tab so 2 tabs opened next to each other and visiting the same page have different sessionStorages which I find really annoying and I'm sure it would trick people.

**Reply by stepanjakl** [reply]
I feel you, but that's how Session Storage operates as a native browser functionality.

Possible workaround: https://stackoverflow.com/a/32766809/11035513

**Reply by SimoTod** [reply]
Yeah, to be fair it's the standard behaviour so as long as it's mentioned im the documentation, it should be fine. 

I also wonder if persist could support an expiry date (useful for local storage).
`$persist(false).for('1 day')`

**Reply by stepanjakl** [reply]
> Yeah, to be fair it's the standard behaviour so as long as it's mentioned im the documentation, it should be fine.
> 
> I also wonder if persist could support an expiry date (useful for local storage).
> `$persist(false).for('1 day')`

No, you can only do expiry date with cookies.

"LocalStorage has no expiration time, Data in the LocalStorage persist till the user manually delete it. This is the only difference between LocalStorage and SessionStorage" -- https://www.geeksforgeeks.org/localstorage-and-sessionstorage-web-storage-apis/

**Reply by SimoTod** [reply]
That doesn't matter too much, it's on your implementation. persist can save a json containing a timestamp field and a value field and it can then decide to use the stored value or the default one based on the expiry.

Cookies shouldn't really be used to store info like that imo (personal opinion, though).

**Reply by stepanjakl** [reply]
Ahh, yes, using the timestamp sounds like a good approach 👍

**Comment by SimoTod** [comment]
I played a bit with that idea and we can have `$persist('foo').in(sessionStorage)` with just 17 bytes more.

It would also support any custom storage object exposing getItem and setItem so it's possible to have
For example
```html
    <script>
      window.cookieStorage = {
        getItem(key) {
          let cookies = document.cookie.split(";");
          for (let i = 0; i < cookies.length; i++) {
            let cookie = cookies[i].split("=");
            if (key == cookie[0].trim()) {
                return JSON.parse(decodeURIComponent(cookie[1]));
            }
          }
          return null;
        },
        setItem(key, value) {
          document.cookie = key+' = '+encodeURIComponent(JSON.stringify(value))
        }
      }
    </script>

    <div x-data="{ open: $persist(false).in(cookieStorage) }">
        <button @click="open = !open">Toggle</button>

        <span x-show="open">
            Content...
        </span>
    </div>
 ```
 
 I've just created a PR to fix another issue with $persist. I'll mention this thread so we can see if Caleb would accept the feature.

**Reply by calebporzio** [reply]
Dude, this is freaking brilliant. I love it.

The only question remaining is: Is `.in()` the best name. I think it is, but the other options would be `.into` or `to`.

If I were speaking the code like a sentence, I'd say "to" -> "Persist this variable TO session storage"

How do you feel about "to"

Anyhow, let me say again how I love this in that it allows for extending the behavior with the exact same API that the browser uses so you can easily just use those. Solid stuff. Would gladly accept PR

**Reply by SimoTod** [reply]
Brilliant, I switched to 'to(...)'. 👍

**Reply by ADTC** [reply]
@SimoTod there are some important caveats here:

The space around `=` is not necessary, and probably best avoided: `document.cookie = key+'='+...` or ``document.cookie = `${key}=...``.

It's easier to apply trim after split using map functions:

* `document.cookie.split(';').map((cookie) => cookie.trim())`
* `cookies[i].split('=').map((part) => part.trim())`

## Double quotes for string values in cookies

The `setItem` method gets the value **wrapped in double quotes** because it's a `JSON.stringify` output. If you apply `encodeURIComponent` to it, you'll get those converted to `%22`.

Generally this is safe as the cookie value is a stringified JSON object. But in special cases, maybe your cookie value is just a simple string, and the quote character encoding or the quotes themselves are undesirable.

There are two approaches.

### First approach

If quote encoding is to be avoided, strip and re-add the quotes:

```js
setItem(key: string, value: string) {
  value = value.replace(/^"|"$/g, '');
  document.cookie = `${key}="${encodeURIComponent(value)};"`;
}
```

In this case, `getItem` can remain as `return decodeURIComponent(cookie[1]);`.

### Second approach

If quotes themselves should be avoided, strip the quotes when setting, and add them when getting:

```js
setItem(key: string, value: string) {
  value = value.replace(/^"|"$/g, '');
  document.cookie = `${key}=${encodeURIComponent(value)};`;
}
```

In this case, `getItem` should be:
```js
const value = decodeURIComponent(cookie[1]);
return `"${value}"`;
```
or you'll get `JSON.parse` errors from the plugin.

> [!NOTE]
> Each of the two approaches has a subtle difference: The first approach will keep the quotes in the cookie value. The second approach will remove them. So you'd have to pick which one works better for you.

## A few final things:

The cookie expires on session expiry. You may want a 60-day expiry instead (for example):

```js
// 60 days expiration for this cookie
const expiry = new Date(Date.now() + 60 * 24 * 60 * 60 * 1000).toUTCString()
document.cookie = `${key}=${encodeURIComponent(value)}; expires=${expiry};`
```

You may want to set a fixed path for the cookie, instead of the current path:

```js
const path = 'path=/;'
document.cookie = `${key}=${encodeURIComponent(value)}; ${path}`
```

You may want to delete the cookie on empty value instead of setting an empty cookie:

```js
if (value === `""`) document.cookie = `${key}=${encodeURIComponent(value)}; expires=${expiry}; ${path}`
else document.cookie = `${key}=; expires=Thu, 01 Jan 1970 00:00:00 UTC; ${path}`
```

Note: If you stripped the double quotes, ``if (value === `""`)`` should be `if (value)` instead.

## Issue: The name of the cookie has `_x_` prefixed to it.

To avoid this, the usage should be something like `open: $persist(false).as('open').in(cookieStorage)`

**Comment by robdekort** [comment]
Just found out about .using(sessionStorage). This is brilliant for this "annoyance bar" I had to add to a site. Thanks @SimoTod.
